File: macos/Sources/RewordMeData/RewordService.swift
import Foundation
import RewordMeDomain
import RewordMeModels

/// Performs the actual HTTP calls. The wire formats live in the injected
/// provider clients; this type only owns transport and error mapping.
public struct RewordService: Sendable {
    private let session: URLSession
    private let registry: ProviderClientRegistry
    private let accountProviders: AccountProviderService
    private let localModel: LocalModelManager

    public init(
        session: URLSession = .shared,
        registry: ProviderClientRegistry = ProviderClientRegistry(),
        accountProviders: AccountProviderService = AccountProviderService(),
        localModel: LocalModelManager = LocalModelManager()
    ) {
        self.session = session
        self.registry = registry
        self.accountProviders = accountProviders
        self.localModel = localModel
    }

    public func listModels(
        provider: ProviderKind,
        apiKey: String,
        endpoint: URL? = nil
    ) async throws -> [ModelInfo] {
        switch provider {
        case .local:
            return LocalModelCatalog.all.map { ModelInfo(id: $0.id, displayName: $0.displayName) }
        case .codex:
            return [ModelInfo(id: "automatic", displayName: "Automatic (Codex)")]
        case .claudeAccount:
            return [
                ModelInfo(id: "automatic", displayName: "Automatic (Claude)"),
                ModelInfo(id: "sonnet", displayName: "Sonnet"),
                ModelInfo(id: "opus", displayName: "Opus")
            ]
        default:
            break
        }
        let key = try validated(apiKey, for: provider)
        let client = registry.client(for: provider)
        let data = try await perform(client.modelsRequest(apiKey: key, endpoint: endpoint))
        return try client.parseModels(data)
    }

    public func reword(
        provider: ProviderKind,
        apiKey: String,
        model: String,
        systemPrompt: String,
        text: String,
        endpoint: URL? = nil
    ) async throws -> String {
        if provider.isAccountProvider {
            return try await accountProviders.reword(
                provider: provider,
                model: model,
                systemPrompt: systemPrompt,
                text: text
            )
        }
        if provider == .local {
            let manifest = LocalModelCatalog.model(id: model)
            let connection = try await localModel.connection(modelID: manifest.id)
            let client = registry.client(for: .local)
            let request = try client.rewordRequest(
                apiKey: connection.apiKey,
                model: manifest.id,
                systemPrompt: systemPrompt,
                text: text,
                endpoint: connection.endpoint
            )
            return try client.parseReword(try await perform(request))
        }
        let key = try validated(apiKey, for: provider)
        let client = registry.client(for: provider)
        let request = try client.rewordRequest(
            apiKey: key,
            model: model,
            systemPrompt: systemPrompt,
            text: text,
            endpoint: endpoint
        )
        let data = try await perform(request)
        return try client.parseReword(data)
    }

    private func validated(_ apiKey: String, for provider: ProviderKind) throws -> String {
        guard provider.requiresAPIKey else {
            // Local services ignore this placeholder; managed/account paths
            // are intercepted before reaching validation.
            return provider == .ollama ? "ollama" : "local"
        }
        let trimmed = apiKey.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { throw RewordError.missingAPIKey }
        return trimmed
    }

    private func perform(_ request: URLRequest) async throws -> Data {
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw RewordError.invalidResponse
        }
        switch http.statusCode {
        case 200...299:
            return data
        case 401, 403:
            throw RewordError.invalidAPIKey
        case 429:
            if Self.isUsageLimitError(data) {
                throw RewordError.usageLimitReached
            }
            let retryAfter = http.value(forHTTPHeaderField: "retry-after").flatMap { Int($0) }
            throw RewordError.rateLimited(retryAfterSeconds: retryAfter)
        default:
            throw RewordError.apiError(
                status: http.statusCode,
                message: Self.errorMessage(from: data)
            )
        }
    }

    /// Best-effort extraction of a human-readable message from any of the
    /// providers' error envelopes.
    static func errorMessage(from data: Data) -> String {
        guard let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return String(data: data.prefix(200), encoding: .utf8) ?? "Unknown error"
        }
        if let error = object["error"] as? [String: Any] {
            if let message = error["message"] as? String { return message }
        }
        if let message = object["message"] as? String { return message }
        return String(data: data.prefix(200), encoding: .utf8) ?? "Unknown error"
    }

    /// HTTP 429 also covers exhausted credits and enforced spend/usage limits.
    /// Those failures require user action and must not be presented as a
    /// temporary rate limit. OpenAI exposes the distinction in `error.code`
    /// while older responses may only carry `error.type` or a message.
    static func isUsageLimitError(_ data: Data) -> Bool {
        guard let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return false
        }
        let error = object["error"] as? [String: Any]
        let fields = [
            error?["code"] as? String,
            error?["type"] as? String,
            object["code"] as? String,
            object["type"] as? String
        ]
        let actionRequiredCodes: Set<String> = [
            "credit_balance_exhausted",
            "organization_spend_limit_exceeded",
            "project_spend_limit_exceeded",
            "organization_usage_limit_exceeded",
            "insufficient_quota",
            "billing_hard_limit_reached",
            "billing_not_active"
        ]
        if fields.compactMap({ $0?.lowercased() }).contains(where: actionRequiredCodes.contains) {
            return true
        }

        let message = errorMessage(from: data).lowercased()
        let transientMarkers = [
            "rate_limit_error", "rate_limit_exceeded", "resource_exhausted",
            "per minute", "per second", "requests per", "tokens per",
            "rpm", "tpm", "retry after", "retry shortly", "quota metric", "rate limit"
        ]
        if transientMarkers.contains(where: message.contains) { return false }

        let actionRequiredPhrases = [
            "credit balance", "spend limit", "insufficient quota",
            "billing not active", "billing hard limit", "payment required",
            "credits exhausted", "no credits"
        ]
        if actionRequiredPhrases.contains(where: message.contains) { return true }
        return message.contains("current quota") && (
            message.contains("billing") || message.contains("plan") || message.contains("credit")
        )
    }
}

extension RewordService: ModelListing {}

File: kmp/core/data/src/jvmMain/kotlin/dev/mjablonski/rewordme/data/RewordService.kt
package dev.mjablonski.rewordme.data

import dev.mjablonski.rewordme.domain.Rewording
import dev.mjablonski.rewordme.models.ModelInfo
import dev.mjablonski.rewordme.models.LocalModelCatalog
import dev.mjablonski.rewordme.models.ProviderKind
import dev.mjablonski.rewordme.models.RewordError
import io.ktor.client.HttpClient
import io.ktor.client.engine.cio.CIO
import io.ktor.client.request.header
import io.ktor.client.request.request
import io.ktor.client.request.setBody
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.HttpMethod
import io.ktor.http.contentType
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * Performs the actual HTTP calls. Request building and parsing live in
 * the per-provider clients so they stay pure and testable.
 */
class RewordService(
    private val http: HttpClient = HttpClient(CIO),
    private val registry: ProviderClientRegistry = ProviderClientRegistry(),
    private val accountProviders: AccountProviderService = AccountProviderService(),
    private val localModel: LocalModelManager = LocalModelManager()
) : Rewording {
    override suspend fun listModels(
        provider: ProviderKind,
        apiKey: String,
        endpoint: String?
    ): List<ModelInfo> {
        when (provider) {
            ProviderKind.LOCAL -> return LocalModelCatalog.ALL.map { ModelInfo(it.id, it.displayName) }
            ProviderKind.CODEX -> return listOf(ModelInfo("automatic", "Automatic (Codex)"))
            ProviderKind.CLAUDE_ACCOUNT -> return listOf(
                ModelInfo("automatic", "Automatic (Claude)"),
                ModelInfo("sonnet", "Sonnet"),
                ModelInfo("opus", "Opus")
            )
            else -> Unit
        }
        val client = registry.client(provider)
        val body = perform(client.modelsRequest(validated(apiKey, provider), endpoint))
        return client.parseModels(body)
    }

    override suspend fun reword(
        provider: ProviderKind,
        apiKey: String,
        model: String,
        systemPrompt: String,
        text: String,
        endpoint: String?
    ): String {
        if (provider.isAccountProvider) {
            return accountProviders.reword(provider, model, systemPrompt, text)
        }
        if (provider == ProviderKind.LOCAL) {
            val manifest = LocalModelCatalog.model(model)
            val connection = localModel.connection(manifest.id)
            val client = registry.client(ProviderKind.LOCAL)
            val request = client.rewordRequest(
                connection.apiKey, manifest.id, systemPrompt, text, connection.endpoint
            )
            return client.parseReword(perform(request))
        }
        val client = registry.client(provider)
        val request = client.rewordRequest(validated(apiKey, provider), model, systemPrompt, text, endpoint)
        return client.parseReword(perform(request))
    }

    private fun validated(apiKey: String, provider: ProviderKind): String {
        // Managed/account paths are intercepted above; Ollama ignores auth.
        if (!provider.requiresApiKey) return if (provider == ProviderKind.OLLAMA) "ollama" else "local"
        return apiKey.trim().ifEmpty { throw RewordError.MissingApiKey }
    }

    private suspend fun perform(request: ProviderRequest): String {
        val response = http.request(request.url) {
            method = HttpMethod.parse(request.method)
            request.headers.forEach { (name, value) -> header(name, value) }
            request.jsonBody?.let {
                contentType(ContentType.Application.Json)
                setBody(it)
            }
        }
        val body = response.bodyAsText()
        return when (response.status.value) {
            in 200..299 -> body
            401, 403 -> throw RewordError.InvalidApiKey
            429 -> {
                if (isUsageLimitError(body)) throw RewordError.UsageLimitReached
                throw RewordError.RateLimited(response.headers["retry-after"]?.toIntOrNull())
            }
            else -> throw RewordError.Api(response.status.value, errorDetail(body))
        }
    }

    companion object {
        /** Best-effort extraction of a message from any error envelope. */
        internal fun errorDetail(body: String): String {
            val root = runCatching { lenientJson.parseToJsonElement(body).jsonObject }
                .getOrNull() ?: return body.take(200)
            val error = (root["error"] as? JsonObject)
            val message = error?.get("message") ?: root["message"]
            return runCatching { message?.jsonPrimitive?.content }.getOrNull()
                ?: body.take(200)
        }

        /** Distinguishes user-action billing/quota 429s from temporary throttling. */
        internal fun isUsageLimitError(body: String): Boolean {
            val root = runCatching { lenientJson.parseToJsonElement(body).jsonObject }
                .getOrNull() ?: return false
            val error = root["error"] as? JsonObject
            val fields = listOf(
                error?.get("code"),
                error?.get("type"),
                root["code"],
                root["type"]
            ).mapNotNull { value ->
                runCatching { value?.jsonPrimitive?.content?.lowercase() }.getOrNull()
            }
            val actionRequiredCodes = setOf(
                "credit_balance_exhausted",
                "organization_spend_limit_exceeded",
                "project_spend_limit_exceeded",
                "organization_usage_limit_exceeded",
                "insufficient_quota",
                "billing_hard_limit_reached",
                "billing_not_active"
            )
            if (fields.any(actionRequiredCodes::contains)) return true

            val message = errorDetail(body).lowercase()
            val transientMarkers = listOf(
                "rate_limit_error", "rate_limit_exceeded", "resource_exhausted",
                "per minute", "per second", "requests per", "tokens per",
                "rpm", "tpm", "retry after", "retry shortly", "quota metric", "rate limit"
            )
            if (transientMarkers.any(message::contains)) return false

            val actionRequiredPhrases = listOf(
                "credit balance", "spend limit", "insufficient quota",
                "billing not active", "billing hard limit", "payment required",
                "credits exhausted", "no credits"
            )
            if (actionRequiredPhrases.any(message::contains)) return true
            return "current quota" in message &&
                listOf("billing", "plan", "credit").any(message::contains)
        }
    }
}
